tensorflow/core/graph/mkl_layout_pass.cc

```diff
@@ -3401,16 +3674,110 @@ Status MklLayoutRewritePass::RewriteNode(std::unique_ptr<Graph>* g,
   for (const Edge* e : orig_node->out_edges()) {
     if (e->IsControlEdge()) {
-      // Allow duplicate while adding control edge as it would fail (return
-      // NULL) if we try to add duplicate edge.
-      CHECK_NOTNULL((*g)->AddControlEdge(new_node, e->dst(), true));
+      auto result = unique_node.insert(e->dst());
+      if (result.second) {
+        (*g)->AddControlEdge(*new_node, e->dst(), true);
+      }
     } else {
       CHECK_NOTNULL((*g)->AddEdge(
-          new_node,
+          *new_node,
           GetTensorDataIndex(e->src_output(), e->src()->num_outputs()),
           e->dst(), e->dst_input()));
     }
   }
+  return Status::OK();
+}
+
+Status MklLayoutRewritePass::RewriteNodeForJustOpNameChange(
+    std::unique_ptr<Graph>* g, const Node* orig_node, Node** new_node,
+    const RewriteInfo* ri) {
+  // Get all data inputs.
+  int num_data_inputs = orig_node->in_edges().size();
+  // Drop count for control edges from inputs
+  for (const Edge* e : orig_node->in_edges()) {
+    if (e->IsControlEdge()) {
+      num_data_inputs--;
+    }
+  }
+  gtl::InlinedVector<Node*, 4> control_edges;
+  gtl::InlinedVector<std::pair<Node*, int>, 4> inputs(num_data_inputs);
+  FillInputs(orig_node, &control_edges, &inputs);
+
+  // Build new node. We use same name as original node, but change the op name.
+  NodeBuilder nb(orig_node->name().c_str(), ri->new_name.c_str());
+  // Copy user-specified device assigned to original node to new node.
+  nb.Device(orig_node->def().device());
+
+  Status s = CopyInputs(orig_node, inputs, &nb);
+  if (s != Status::OK()) {
+    return s;
+  }
+
+  ri->copy_attrs(const_cast<const Node*>(orig_node), &nb, true);
+  nb.Attr("_kernel", mkl_op_registry::kMklNameChangeOpLabel);
+
+  // Finalize graph and get new node.
+  s = nb.Finalize(&**g, new_node);
+  if (s != Status::OK()) {
+    return s;
+  }
+  CHECK_NOTNULL(*new_node);
+
+  // In the following code of this function, an unsorted set is used to make
+  // sure no duplicated edges be added into the new node. Therefore, we can
+  // pass allow_duplicates = true in AddControlEdge call to skip the O(#edges)
+  // check in the routine.
+
+  // Incoming data edges from 'orig_node' node to new 'new_node' node are
+  // already copied in BuildNode. We need to handle control edges now.
+  std::unordered_set<Node*> unique_node;
+  for (const Edge* e : orig_node->in_edges()) {
+    if (e->IsControlEdge()) {
+      auto result = unique_node.insert(e->src());
+      if (result.second) {
+        (*g)->AddControlEdge(e->src(), *new_node, true);
+      }
+    }
+  }
+  unique_node.clear();
+
+  // Transfer outgoing edges from 'orig_node' node to new 'new_node' node.
+  for (const Edge* e : orig_node->out_edges()) {
+    if (e->IsControlEdge()) {
+      auto result = unique_node.insert(e->dst());
+      if (result.second) {
+        (*g)->AddControlEdge(*new_node, e->dst(), true);
+      }
+    } else {
+      CHECK_NOTNULL((*g)->AddEdge(*new_node, e->src_output(),
```

Comment:
Sorry that I just noticed this. We don't use `CHECK` in new code anymore. Please replace all new `CHECK`s in this PR with `DCHECK`s. 

Note that `DCHECK` turns into a no-op outside debug mode so you shouldn't rely on executing anything inside `DCHECK`. For example:
```
// Wrong approach: giving unexpected behavior
DCHECK((*g)->AddEdge(...));  // AddEdge will not be executed in non-debug mode.

// Correct approach:
auto result = (*g)->AddEdge(...);
DCHECK(result);
```

Reply:
what about CHECK_NOT_NULL()?

Reply:
Any CHECK* should be replaced with DCHECK*. The existing ones are fine, so you only need to change the new CHECK* you use in this PR.

Reply:
There is no definition of DCHECK_NOTNULL like other CHECKs. And I also need to change the TF_CHECK_OK to TF_DCHECK_OK, right?

Reply:
Ah. I see what you mean. For `CHECK_NOTNUL(ptr)`, please replace it with `DCHECK(ptr)` or `DCHECK(ptr != nullptr)`. `TF_CHECK_OK` is fine. Please only change commands that begin with `CHECK`.

Reply:
ok, got it.
have changed.